pandas/tests/test_base.py

```diff
@@ -1175,3 +1176,94 @@ def test_iter_box(self):
             assert isinstance(res, pd.Period)
             assert res.freq == 'M'
             assert res == exp
+
+
+@pytest.mark.parametrize('arr, expected', [
+    (pd.DatetimeIndex(['2017', '2017']), pd.DatetimeIndex(['2017'])),
+    (pd.DatetimeIndex(['2017', '2017'], tz='US/Eastern'),
+     pd.DatetimeIndex(['2017'], tz='US/Eastern')),
+])
+def test_unique_datetime_index(arr, expected):
+    result = arr.unique()
+
+    if isinstance(expected, np.ndarray):
+        tm.assert_numpy_array_equal(result, expected)
+    if isinstance(expected, pd.Series):
+        tm.assert_series_equal(result, expected)
+    if isinstance(expected, pd.DatetimeIndex):
+        tm.assert_index_equal(result, expected)
+
+
+@pytest.mark.parametrize('arr, expected', [
+    (pd.Series(pd.DatetimeIndex(['2017', '2017'])),
+     np.array(['2017'], dtype='M8[ns]')),
+    (pd.Series(pd.DatetimeIndex(['2017', '2017'], tz='US/Eastern')),
+     np.array([pd.Timestamp('2017', tz="US/Eastern")], dtype=object)),
+])
+def test_unique_datetime_series(arr, expected):
+    result = arr.unique()
+
+    if isinstance(expected, np.ndarray):
+        tm.assert_numpy_array_equal(result, expected)
+    if isinstance(expected, pd.Series):
+        tm.assert_series_equal(result, expected)
+    if isinstance(expected, pd.DatetimeIndex):
+        tm.assert_index_equal(result, expected)
+
+
+@pytest.mark.parametrize('array, expected_type, dtype', [
+    (np.array([0, 1]), np.ndarray, 'int64'),
+    (np.array(['a', 'b']), np.ndarray, 'object'),
+    (pd.Categorical(['a', 'b']), pd.Categorical, 'category'),
+    (pd.DatetimeIndex(['2017', '2018']), np.ndarray, 'datetime64[ns]'),
+    (pd.DatetimeIndex(['2017', '2018'], tz="US/Central"), pd.DatetimeIndex,
+     'datetime64[ns, US/Central]'),
+    (pd.TimedeltaIndex([10**10]), np.ndarray, 'm8[ns]'),
+    (pd.PeriodIndex([2018, 2019], freq='A'), np.ndarray, 'object'),
+    (pd.IntervalIndex.from_breaks([0, 1, 2]), np.ndarray, 'object'),
+])
+def test_values_consistent(array, expected_type, dtype):
+    l_values = pd.Series(array)._values
+    r_values = pd.Index(array)._values
+    assert type(l_values) is expected_type
+    assert type(l_values) is type(r_values)
+
+    if isinstance(l_values, np.ndarray):
+        tm.assert_numpy_array_equal(l_values, r_values)
+    elif isinstance(l_values, pd.Index):
+        tm.assert_index_equal(l_values, r_values)
+    elif pd.api.types.is_categorical(l_values):
+        tm.assert_categorical_equal(l_values, r_values)
```

Comment:
In the future, we should maybe add an `assert_array_equal` that handles the different extension arrays as well, then such a special case should not be needed here?